### PR TITLE
Move connectionset from analyzer and protocols from synthesizer

### PR DESCRIPTION
**Pull request by elazarg**
Reopening #8.

Main differences from the version in the synthesizer:
* Remove mutating methods, replace with convenience constructors: All(), None(), TCPorUDPConnection(), ICMPConnection()
* Remove AllowAll field
* Dedicated StatefulState type
* Use netp.ProtocolString instead of ProtocolStr
* switchSrcDstPortsOnTCP() only actually switch dimensions when they differ
* Sort protocol strings lexicographically on output: ICMP,TCP,UDP
* Rename:
  * ConnectionSet => Set
  * getProtocolStr() => protocolStringFromCode()
  * getDimensionDomain() => entireDimension()
  * AllowAll => IsAll()
  * Intersection() => Intersect()
  * ConnWithStatefulness() => WithStatefulness()
  * getDimensionStr() => getDimensionString()
  * Except() => ExceptCidrs()

### CODE REVIEW

**Review by adisos — COMMENTED** [review]
Looks good

**Review by zivnevo — CHANGES_REQUESTED** [review]
We need the copyright comment at the top for all source files

File: pkg/connection/connectionset.go [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+	icmpTypeMin, icmpTypeMax, icmpCodeMin, icmpCodeMax int64) {
+	code := ProtocolStringToCode(protocol)
+	cube := hypercube.FromCubeShort(code, code,
+		srcMinP, srcMaxP, dstMinP, dstMaxP,
+		icmpTypeMin, icmpTypeMax, icmpCodeMin, icmpCodeMax)
+	conn.connectionProperties = conn.connectionProperties.Union(cube)
+	conn.Canonicalize()
+}
+
+func (conn *Set) Canonicalize() {
```
Comment (adisos) [inline pkg/connection/connectionset.go:207]: should `Canonicalize` be exported?

**Review by adisos — COMMENTED** [review]
Looks good!
Added just a small comment for now

File: pkg/connection/connectionset.go (hunk 1 above)
Comment (elazarg) [reply pkg/connection/connectionset.go:207]: No.

File: pkg/ipblock/ipblock.go [hunk 2]
```diff
@@ -209,42 +210,31 @@ func NewIPBlockFromCidrOrAddress(s string) (*IPBlock, error) {
...
-	return res, nil
+	return &IPBlock{ipRange: *ipRange}, nil
 }
 
-// NewIPBlock returns an IPBlock object from input cidr str an exceptions cidr str
-func NewIPBlock(cidr string, exceptions []string) (*IPBlock, error) {
-	res := IPBlock{ipRange: interval.CanonicalSet{}}
-	span, err := cidrToInterval(cidr)
-	if err != nil {
-		return nil, err
```
Comment (adisos) [inline pkg/ipblock/ipblock.go]: I would keep this function (`func NewIPBlock(cidr string, exceptions []string)`)
It is used by another repo, also planned to import this pkg now

File: pkg/ipblock/ipblock_test.go [hunk 3]
```diff
@@ -23,10 +23,6 @@ func TestOps(t *testing.T) {
 	minus := ipb1.Subtract(ipb2)
 	require.Equal(t, "1.2.3.0-1.2.3.3, 1.2.3.5-1.2.3.255", minus.ToIPRanges())
 
-	minus2, err := ipblock.NewIPBlock(ipb1.ToCidrListString(), []string{ipb2.ToCidrListString()})
```
Comment (adisos) [inline pkg/ipblock/ipblock_test.go]: and will also keep these tests

File: pkg/ipblock/ipblock.go (hunk 2 above)
Comment (elazarg) [reply pkg/ipblock/ipblock.go]: I could not find the usage

Comment (elazarg) [reply pkg/ipblock/ipblock.go]: We can add `ExceptCidrs()` function instead:
```go
// ExceptCidrs returns a new IPBlock with all cidr ranges removed 
func (b *IPBlock) ExceptCidrs(cidrExceptions ...string) (*IPBlock, error) {
	res := b.Copy()
	for i := range cidrExceptions {
		hole, err := NewIPBlockFromCidr(cidrExceptions[i])
		if err != nil {
			return nil, err
		}
		res = res.Subtract(hole)
	}
	return res, nil
}
```

Comment (adisos) [reply pkg/ipblock/ipblock.go]: > I could not find the usage

It's in another repo (not vpc analysis/snyhtesis) that will also import models , for example:
https://github.com/np-guard/netpol-analyzer/blob/main/pkg/netpol/eval/internal/k8s/netpol.go#L318

Comment (elazarg) [reply pkg/ipblock/ipblock.go]: Ok, good to know... I will add it to my workspace and see.

File: pkg/connection/connectionset.go [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+// ContainedIn returns true if c is subset of other
+func (c *Set) ContainedIn(other *Set) bool {
+	res, err := c.connectionProperties.ContainedIn(other.connectionProperties)
+	if err != nil {
+		log.Fatalf("invalid connection set. %e", err)
+	}
+	return res
+}
+
+func ProtocolStringToCode(protocol netp.ProtocolString) int64 {
```
Comment (adisos) [inline pkg/connection/connectionset.go:153]: should not be exported?

**Review by adisos — COMMENTED** [review]
looks good! just one small comment

File: pkg/connection/connectionset.go (hunk 4 above)
Comment (elazarg) [reply pkg/connection/connectionset.go]: fixed

**Review by adisos — APPROVED** [review]

**Review by zivnevo — CHANGES_REQUESTED** [review]

**Comment by elazarg** [comment]
I realized that the dimension types encode the order of the dimensions, which makes it harder to know what depends on the order of the dimensions, and contradicts the comment preceding `dimensionsList`:
> // this should be the only place where the order is hard-coded

So the last commit changes the type of Dimension from int to string; the index of each dimension in a cube is looked up in the slice `dimensionsList`. It also delegates most of the work of swapping dimensions to the underlying hypercube set.

File: pkg/interval/intervalset.go [hunk 5]
```diff
@@ -175,29 +181,30 @@ func (c *CanonicalSet) Overlap(other *CanonicalSet) bool {
...
-	for _, hole := range other.intervalSet {
-		newIntervalSet := []Interval{}
-		for _, interval := range res {
-			newIntervalSet = append(newIntervalSet, interval.subtract(hole)...)
-		}
-		res = newIntervalSet
-	}
-	return &CanonicalSet{
-		intervalSet: res,
+	res := c.Copy()
```
[the remark this answers is not in the document]
Comment (elazarg) [reply pkg/interval/intervalset.go:187]: Added.

File: pkg/interval/intervalset.go [hunk 6]
```diff
@@ -175,29 +181,30 @@ func (c *CanonicalSet) Overlap(other *CanonicalSet) bool {
...
 
 func (c *CanonicalSet) IsSingleNumber() bool {
-	if len(c.intervalSet) == 1 && c.intervalSet[0].Start == c.intervalSet[0].End {
+	if len(c.intervalSet) == 1 && c.intervalSet[0].Size() == 1 {
 		return true
 	}
 	return false
 }
 
+func (c *CanonicalSet) Elements() []int64 {
```
[the remark this answers is not in the document]
Comment (elazarg) [reply pkg/interval/intervalset.go:203]: Added

File: pkg/connection/connectionset.go [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+		return c.Copy()
+	}
+	return &Set{connectionProperties: c.connectionProperties.Subtract(other.connectionProperties)}
+}
+
+// ContainedIn returns true if c is subset of other
+func (c *Set) ContainedIn(other *Set) bool {
+	res, err := c.connectionProperties.ContainedIn(other.connectionProperties)
+	if err != nil {
+		log.Fatalf("invalid connection set. %e", err)
```
[the remark this answers is not in the document]
Comment (elazarg) [reply pkg/connection/connectionset.go]: Done.

File: pkg/hypercube/hypercubeset_test.go [hunk 8]
```diff
@@ -6,18 +8,13 @@ import (
...
 // the input cube is given as an ordered list of integer values, where each two values
 // represent the range (start,end) for a dimension value
 func cube(values ...int64) *hypercube.CanonicalSet {
-	cube := []*interval.CanonicalSet{}
-	for i := 0; i < len(values); i += 2 {
-		cube = append(cube, interval.New(values[i], values[i+1]).ToSet())
-	}
-	return hypercube.FromCube(cube)
+	return hypercube.Cube(values...)
 }
```
[the remark this answers is not in the document]
Comment (elazarg) [reply pkg/hypercube/hypercubeset_test.go]: It's a bit longer... done.

File: pkg/ipblock/ipblock.go [hunk 9]
```diff
@@ -164,15 +162,13 @@ func DisjointIPBlocks(set1, set2 []*IPBlock) []*IPBlock {
...
 	for _, ipb := range ipbList {
-		blocksWithNoOverlaps = addIntervalToList(ipb, blocksWithNoOverlaps)
+		res = addIntervalToList(ipb, res)
 	}
 
-	res := blocksWithNoOverlaps
 	if len(res) == 0 {
-		newAll := GetCidrAll()
-		res = append(res, newAll)
+		res = append(res, GetCidrAll())
```
[the remark this answers is not in the document]
Comment (elazarg) [reply pkg/ipblock/ipblock.go]: Done

File: pkg/interval/intervalset.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (elazarg) [reply pkg/interval/intervalset.go:203]: Also implemented by first allocating all the memory.

File: pkg/hypercube/hypercubeset.go [hunk 10]
```diff
@@ -249,6 +251,24 @@ func (c *CanonicalSet) GetCubesList() [][]*interval.CanonicalSet {
 	return res
 }
 
+// SwapDimensions returns a new CanonicalSet object, built from the input CanonicalSet object,
+// with dimensions dim1 and dim2 swapped
+func (c *CanonicalSet) SwapDimensions(dim1, dim2 int) *CanonicalSet {
```
Comment (zivnevo) [inline pkg/hypercube/hypercubeset.go:256]: Please add unit-test for this function

File: pkg/ipblock/ipblock.go
Comment (zivnevo) [inline pkg/ipblock/ipblock.go:1]: Should we have an (exported?) global var here representing the full range of IP addresses?

Comment (elazarg) [reply pkg/ipblock/ipblock.go:1]: No idea

File: pkg/hypercube/hypercubeset.go (hunk 10 above)
Comment (elazarg) [reply pkg/hypercube/hypercubeset.go:257]: Done

**Pull request merged**